File: scraper.js
import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';

const CONFIG_PATH = './config.json';
const DATA_DIR = './data';

async function scrape() {
  const config = JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf-8'));

  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }

  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  const now = new Date().toISOString().split('T')[0]; // YYYY-MM-DD

  for (const product of config.products) {
    console.log(`Scraping: ${product.name} (${product.site})`);
    console.log(`  URL: ${product.url}`);

    try {
      const page = await browser.newPage();
      await page.setUserAgent(
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
      );
      await page.goto(product.url, { waitUntil: 'networkidle2', timeout: 30000 });

      // Try each selector (comma-separated) until one works
      const selectors = product.selector.split(',').map(s => s.trim());
      let priceText = null;

      for (const sel of selectors) {
        try {
          await page.waitForSelector(sel, { timeout: 5000 });
          priceText = await page.$eval(sel, el => el.textContent.trim());
          if (priceText) {
            console.log(`  Matched selector: ${sel}`);
            break;
          }
        } catch {
          // selector not found, try next
        }
      }

      if (!priceText) {
        // Fallback: look for any element containing a dollar amount on the page
        priceText = await page.evaluate(() => {
          const elements = document.querySelectorAll('*');
          for (const el of elements) {
            if (el.children.length === 0) {
              const text = el.textContent.trim();
              if (/^\$[\d,]+(\.\d{2})?$/.test(text)) {
                return text;
              }
            }
          }
          return null;
        });
        if (priceText) console.log(`  Found price via fallback scan`);
      }

      await page.close();

      if (!priceText) {
        console.log(`  WARNING: No price found for ${product.name}`);
        continue;
      }

      // Clean price: "$1,299.99" -> 1299.99
      const price = parseFloat(priceText.replace(/[^0-9.]/g, ''));
      if (isNaN(price)) {
        console.log(`  WARNING: Could not parse price "${priceText}"`);
        continue;
      }

      console.log(`  Price: $${price}`);

      // Load or create data file
      const dataFile = path.join(DATA_DIR, `${product.id}.json`);
      let data = { id: product.id, name: product.name, url: product.url, site: product.site, history: [] };

      if (fs.existsSync(dataFile)) {
        data = JSON.parse(fs.readFileSync(dataFile, 'utf-8'));
        // Update name/url/site in case config changed
        data.name = product.name;
        data.url = product.url;
        data.site = product.site;
      }

      // Only add if we don't already have today's price
      const lastEntry = data.history[data.history.length - 1];
      if (lastEntry && lastEntry.date === now) {
        console.log(`  Already have today's price, updating...`);
        lastEntry.price = price;
      } else {
        data.history.push({ date: now, price });
      }

      fs.writeFileSync(dataFile, JSON.stringify(data, null, 2));
      console.log(`  Saved to ${dataFile}`);

    } catch (err) {
      console.error(`  ERROR scraping ${product.name}: ${err.message}`);
    }
  }

  await browser.close();
  console.log('\nDone.');
}

scrape();

File: scraper.py
import random
import re
import time
from playwright.sync_api import sync_playwright
from playwright_stealth import Stealth
from config import STORE_CONFIGS, detect_store

stealth = Stealth()


def scrape_all(products, skip_cloudflare=False):
    """Scrape prices for all products. Returns list of (product_id, price) tuples."""
    results = []
    errors = []

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
            viewport={"width": 1920, "height": 1080},
        )
        stealth.apply_stealth_sync(context)

        for i, product in enumerate(products):
            store = detect_store(product["url"])
            if not store:
                errors.append((product["id"], f"Unknown store for URL: {product['url']}"))
                continue

            cfg = STORE_CONFIGS[store]

            if skip_cloudflare and cfg.get("cloudflare"):
                print(f"  Skipping {product['name']} (Cloudflare, --skip-cloudflare)")
                continue

            print(f"  Scraping: {product['name']}")
            print(f"    URL: {product['url']}")

            try:
                page = context.new_page()

                page.goto(product["url"], wait_until="networkidle", timeout=30000)

                # Extra wait for Cloudflare-protected sites
                if cfg.get("cloudflare"):
                    time.sleep(random.uniform(2, 4))

                # Try the configured selector
                price_text = None
                try:
                    page.wait_for_selector(cfg["wait_for"], timeout=10000)
                    el = page.query_selector(cfg["price_selector"])
                    if el:
                        price_text = el.text_content().strip()
                        print(f"    Matched selector: {cfg['price_selector']}")
                except Exception:
                    pass

                # Fallback: scan page for dollar amounts
                if not price_text:
                    price_text = page.evaluate("""() => {
                        const els = document.querySelectorAll('*');
                        for (const el of els) {
                            if (el.children.length === 0) {
                                const t = el.textContent.trim();
                                if (/^\\$[\\d,]+(\\.[\\d]{2})?$/.test(t)) return t;
                            }
                        }
                        return null;
                    }""")
                    if price_text:
                        print(f"    Found price via fallback scan")

                page.close()

                if not price_text:
                    errors.append((product["id"], "No price found on page"))
                    print(f"    WARNING: No price found")
                    continue

                # Parse: "$1,299.99" -> 1299.99
                price = float(re.sub(r"[^\d.]", "", price_text))
                print(f"    Price: ${price:.2f}")
                results.append((product["id"], price))

            except Exception as e:
                errors.append((product["id"], str(e)))
                print(f"    ERROR: {e}")

            # Polite delay between products
            if i < len(products) - 1:
                delay = random.uniform(3, 5)
                time.sleep(delay)

        browser.close()

    if errors:
        print(f"\n  {len(errors)} error(s):")
        for pid, msg in errors:
            print(f"    {pid}: {msg}")

    return results


def fetch_product_name(url):
    """Fetch the product name from a URL by scraping the page title."""
    store = detect_store(url)
    if not store:
        return None

    cfg = STORE_CONFIGS[store]

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        )
        stealth.apply_stealth_sync(context)
        page = context.new_page()

        page.goto(url, wait_until="networkidle", timeout=30000)

        if cfg.get("cloudflare"):
            time.sleep(random.uniform(2, 4))

        name = None
        try:
            el = page.query_selector(cfg["name_selector"])
            if el:
                name = el.text_content().strip()
        except Exception:
            pass

        if not name:
            name = page.title().split("|")[0].split("-")[0].strip()

        browser.close()
        return name
